### PR TITLE
Guidelines pre-PEP: add One header, Naming, Language support

### PR DESCRIPTION
**Pull request by encukou**
Hello,
Here are a few more proposed sections for the upcoming guidelines PEP.

@zooba @vstinner @erlend-aasland @serhiy-storchaka @mdboom, please add your comments or approvals.

I'd like to avoid growing the scope of this PR. For substantial additions, or things that need more discussion, there's a TODO list in #44.
If a section is controversial I can delete it for now.

### CODE REVIEW

File: guidelines.rst [hunk 1]
```diff
@@ -154,6 +154,134 @@ The more stable the tier is, the stricter the rules for it are:
...
+Do not reuse names
+------------------
+
+If an API's behavior changes in a backwards-incompatible way,
+for example when a function's signature changes,
+add new API with a new name.
+You can deprecate and remove the old version following Python's
+:pep:`backwards compatibility policy <387>`.
+
+After API has been removed, do not reuse the old name,
```
Comment (serhiy-storchaka) [inline guidelines.rst:224]: It could be helpful to have a list of "retired" names to check for conflicts.

**Review by serhiy-storchaka — APPROVED** [review]
In general LGTM, but I think that the `PY_` prefix should be used in some cases.

File: guidelines.rst [hunk 2]
```diff
@@ -154,6 +154,134 @@ The more stable the tier is, the stricter the rules for it are:
...
+.. XXX add a PEP number prefix to the anchor:
+
+.. _naming:
+
+Naming
+======
+
+[TODO: PEP 7 should be updated to link here once this PEP goes live.]
+
+All public names must be prefixed with ``Py``.
```
[the remark this answers is not in the document]
Comment (encukou) [reply guidelines.rst:183]: [PEP 7](https://peps.python.org/pep-0007/#naming-conventions) explicitly says that *macros should have a MixedCase prefix*, and I see no reason to change that decision.

File: guidelines.rst (hunk 1 above)
Comment (encukou) [reply guidelines.rst:224]: Yes. On the other hand,
- I don't think it's too likely that someone reuses a name accidentally, and
- if they do, it's a big problem.

The main point is that you shouldn't intentionally retire API to make its name available.

**Comment by erlend-aasland** [comment]
Looks good; I added a few minor remarks.

File: guidelines.rst [hunk 3]
```diff
@@ -154,6 +154,134 @@ The more stable the tier is, the stricter the rules for it are:
...
+contact the C API working group.
+
+The ``Py_`` prefix is in mixed case, even in macro names.
+For example: ``PyUnicode_AS_STRING``.
+(Several existing macros use the upper-case ``PY``; if you need this prefix
+for consistency, please get approval from the C API working group.)
+
+Unstable API is prefixed with ``PyUnstable_`` instead of ``Py``,
+for example ``PyUnstable_Long_IsCompact`` or (hypothetically)
+``PyUnstable_String_GET_SIZE``.
```
[the remark this answers is not in the document]
Comment (encukou) [reply guidelines.rst:206]: There's no good place to link to. PEP 689 is a historical document introducing unstable API, and the devguide is a mutable collection of best practices.

The link should be to a section in this PEP, or another Process PEP, but we don't have it yet.

File: guidelines.rst [hunk 4]
```diff
@@ -154,6 +154,134 @@ The more stable the tier is, the stricter the rules for it are:
...
+
+It is OK to use other features -- compiler-specific ones,
+optional standard ones, or platform-specific ones -- if the behavior of
+correct user code is the same as with a standard compiler.
+For example, compiler-specific code is often used to improve performance
+or compiler diagnostics.
+
+It is also OK to use these other features for platform-specific API,
+which needs to be documented as such and have a feature test macro
+(for example, ``PY_HAVE_THREAD_NATIVE_ID``).
```
Comment (vstinner) [inline guidelines.rst:263]: PY_ prefix is a bad example :-) Maybe use Py_HAVE_C_COMPLEX example.

**Review by vstinner — APPROVED** [review]
LGTM overall, I just left minor comments.

File: guidelines.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (encukou) [reply guidelines.rst:183]: It's a few lines down:
```
(Several existing macros use the upper-case ``PY``; if you need this prefix
for consistency, please get approval from the C API Working Group.)
```

IMO, it's better to cover `Py`, `_Py`, `Py_`, and *then* get to `PY`, which you should only use in exceptional cases. The whole PEP is about *new* API, after all.

File: guidelines.rst [hunk 5]
```diff
@@ -154,6 +154,134 @@ The more stable the tier is, the stricter the rules for it are:
...
+
+Naming
+======
+
+[TODO: PEP 7 should be updated to link here once this PEP goes live.]
+
+All public names must be prefixed with ``Py``.
+
+Names that users should not use directly, but need to be visible to the
+compiler/linker, should be prefixed with ``_Py``.
```
[the remark this answers is not in the document]
Comment (encukou) [reply guidelines.rst:186]: How would you define, for example, `Py_TYPE` without using `_Py_TYPE`?

File: guidelines.rst (hunk 4 above)
Comment (encukou) [reply guidelines.rst:263]: IMO, undocumented API is worse than an older naming style.

File: guidelines.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (encukou) [reply guidelines.rst:183]: But ke can make *this* clear:
```suggestion
All newly added public names must be prefixed with ``Py``.
```

File: guidelines.rst (hunk 4 above)
Comment (erlend-aasland) [reply guidelines.rst:263]: > IMO, undocumented API is a worse example than one with an older naming style.

`PY_HAVE_THREAD_NATIVE_ID` is not documented.

File: guidelines.rst [hunk 6]
```diff
@@ -154,6 +154,134 @@ The more stable the tier is, the stricter the rules for it are:
...
+To allow selecting an alternate API, such as a subset,
+use feature flags/macros that users define before including the header
+(for example ``Py_LIMITED_API``).
+Adding such a feature macro needs approval from the C API Working Group.
+
+.. note::
+
+    For background and discussions, see:
+
+    - https://github.com/capi-workgroup/api-evolution/issues/34
```
Comment (zooba) [inline guidelines.rst:171]: Don't these need to be 3 space indented to match the `.. note`? (I'm not sure if that's a real ReST thing or not, but I learned to do it that way at some point, which I assume means a compiler shouted at me)

File: guidelines.rst [hunk 7]
```diff
@@ -154,6 +154,134 @@ The more stable the tier is, the stricter the rules for it are:
...
+
+    - https://github.com/capi-workgroup/api-evolution/issues/21
+
+
+Do not reuse names
+------------------
+
+If an API's behavior changes in a backwards-incompatible way,
+for example when a function's signature changes,
+add new API with a new name.
```
Comment (zooba) [inline guidelines.rst:220]: Agreed, but "behavior" and "function signature" are pretty different things. Can we say "interface or behavior"? Or maybe use borrowed/stolen reference changes as the example?

File: guidelines.rst [hunk 8]
```diff
@@ -154,6 +154,134 @@ The more stable the tier is, the stricter the rules for it are:
...
+  - ``static inline`` functions
+  - designated initializers
+  - intermingled declarations
+  - line comments (``//``)
+
+- C++03
+
+It is OK to use other features -- compiler-specific ones,
+optional standard ones, or platform-specific ones -- if the behavior of
+correct user code is the same as with a standard compiler.
```
Comment (zooba) [inline guidelines.rst:257]: Something like "compiler-specific features should be detected using appropriate preprocessor checks, and their use should not produce warnings on any supported compiler, including earlier versions of the one it is specific to"?

File: guidelines.rst [hunk 9]
```diff
@@ -154,6 +154,134 @@ The more stable the tier is, the stricter the rules for it are:
...
+   The existing API uses a few C11 features which are
+   commonly available as compiler extensions to C99.
+   In particular, we do use an unnamed union.
+   New API should not use these features.
+
+All function declarations and definitions must use full prototypes,
+that is, the types of all arguments must be specified.
+
+Argument *names* should generally be included in declarations,
+but may be omitted if the meaning is obvious.
```
Comment (zooba) [inline guidelines.rst:276]: Do we gain anything from this? Might as well just require it, IMHO. And should we require that names match the definition? (I suppose allowing variation saves us from PRs making fixes to our existing inconsistencies...)

**Review by zooba — COMMENTED** [review]
LGTM. Would like to see a little bit more about the compiler-specific features comment, but otherwise happy to see this merge.

File: guidelines.rst (hunk 8 above)
Comment (encukou) [reply guidelines.rst]: OK, but that would be a good guideline for C code in general, not just the headers/API.

File: guidelines.rst (hunk 9 above)
Comment (zooba) [reply guidelines.rst:280]: > I don't plan PRs to fix inconsistencies.

I know _you_ don't, but we'll get other people who will come along and create PRs to fix things that don't match new guidelines.

Maybe it's best to just say nothing?

Comment (encukou) [reply guidelines.rst]: OK. PEP-7 would be a better place for this, anyway.

File: guidelines.rst (hunk 4 above)
Comment (encukou) [reply guidelines.rst]: Silly me.

**Comment by encukou** [comment]
Thank you for the reviews, everyone!
I plan to merge on Wednesday, leaving more changes for follow-up PRs.

**Pull request merged**